hazelcast/src/main/java/com/hazelcast/map/impl/MapKeyLoader.java

```diff
@@ -0,0 +1,297 @@
+
+        while (batches.hasNext()) {
+            Map<Integer, List<Data>> batch = batches.next();
+            sendBatch(batch, replaceExistingValues);
+        }
+
+        sendLoadCompleted(partitionService.getPartitionCount(), replaceExistingValues);
+
+        if (keys instanceof Closeable) {
+            closeResource((Closeable) keys);
+        }
+    }
+
+    public static Iterator<Map<Integer, List<Data>>> toBatches(final Iterator<Entry<Integer, Data>> entries,
+            final int maxBatch) {
+
+        return new UnmodifiableIterator<Map<Integer, List<Data>>>() {
+            @Override
+            public boolean hasNext() {
+                return entries.hasNext();
+            }
+
+            @Override
+            public Map<Integer, List<Data>> next() {
+                ValidationUtil.checkHasNext(entries, "No next element");
+                return nextBatch(entries, maxBatch);
+            }
+        };
+    }
+
+    private IFunction<Data, Entry<Integer, Data>> toPartition() {
+        return new IFunction<Data, Entry<Integer, Data>>() {
+            @Override
+            public Entry<Integer, Data> apply(Data input) {
+                Integer partition = partitionService.getPartitionId(input);
+                return new MapEntrySimple<Integer, Data>(partition, input);
+            }
+        };
+    }
+
+    private static Map<Integer, List<Data>> nextBatch(Iterator<Entry<Integer, Data>> entries, int maxBatch) {
+
+        Map<Integer, List<Data>> batch = new HashMap<Integer, List<Data>>();
+
+        while (entries.hasNext()) {
+            Entry<Integer, Data> e = entries.next();
+            List<Data> partitionKeys = CollectionUtil.addToValueList(batch, e.getKey(), e.getValue());
+
+            if (partitionKeys.size() >= maxBatch) {
+                break;
+            }
+        }
+
+        return batch;
+    }
+
+    private void sendBatch(Map<Integer, List<Data>> batch, boolean replaceExistingValues) {
+
+        for (Entry<Integer, List<Data>> e : batch.entrySet()) {
+            int partitionId = e.getKey();
+            List<Data> keys = e.getValue();
+            LoadAllOperation op = new LoadAllOperation(mapName, keys, replaceExistingValues, false);
+            opService.invokeOnPartition(SERVICE_NAME, op, partitionId);
+        }
+
+    }
+
+    private List<Future<Object>> sendLoadCompleted(int partitions, boolean replaceExistingValues) {
+
+        List<Future<Object>> futures = new ArrayList<Future<Object>>();
+        boolean lastBatch = true;
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            LoadAllOperation op = new LoadAllOperation(mapName, Collections.<Data>emptyList(), replaceExistingValues, lastBatch);
+            futures.add(opService.invokeOnPartition(SERVICE_NAME, op, partitionId));
+        }
+
+        return futures;
+    }
+
+    public static int getMaxSize(int clusterSize, MaxSizeConfig maxSizeConfig) {
```

Comment:
Maybe OOM is not a concern since you are doing batching but it seems load times may increase when using PER_PARTITION max size policy due to that negative return value.